pkg/controller/crdmirroring/crdhandler/clustergroup.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crdhandler
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	crd "github.com/vmware-tanzu/antrea/pkg/apis/crd/v1alpha2"
+	crdclientset "github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned"
+	crdinformer "github.com/vmware-tanzu/antrea/pkg/client/informers/externalversions/crd/v1alpha2"
+	crdlister "github.com/vmware-tanzu/antrea/pkg/client/listers/crd/v1alpha2"
+	"github.com/vmware-tanzu/antrea/pkg/controller/crdmirroring/types"
+	legacycore "github.com/vmware-tanzu/antrea/pkg/legacyapis/core/v1alpha2"
+	legacycrdclientset "github.com/vmware-tanzu/antrea/pkg/legacyclient/clientset/versioned"
+	legacycoreinformer "github.com/vmware-tanzu/antrea/pkg/legacyclient/informers/externalversions/core/v1alpha2"
+	legacycorelister "github.com/vmware-tanzu/antrea/pkg/legacyclient/listers/core/v1alpha2"
+)
+
+type ClusterGroupHandler struct {
+	lister       crdlister.ClusterGroupLister
+	legacyLister legacycorelister.ClusterGroupLister
+	client       crdclientset.Interface
+	legacyClient legacycrdclientset.Interface
+	CRDName      string
+}
+
+func newClusterGroupHandler(informer, legacyInformer types.GenericInformer,
+	client crdclientset.Interface,
+	legacyClient legacycrdclientset.Interface,
+	CRDName string) types.MirroringHandler {
+	mc := &ClusterGroupHandler{
+		lister:       informer.(crdinformer.ClusterGroupInformer).Lister(),
+		legacyLister: legacyInformer.(legacycoreinformer.ClusterGroupInformer).Lister(),
+		client:       client,
+		legacyClient: legacyClient,
+		CRDName:      CRDName,
+	}
+	return mc
+}
+
+// GetNewObject gets the mirroring new ClusterGroup struct
+func (c *ClusterGroupHandler) GetNewObject(namespace, name string) (metav1.Object, error) {
+	lister := c.lister
+	n, err := lister.Get(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, fmt.Errorf("failed to get new %s %s/%s with lister: %v", c.CRDName, namespace, name, err)
+	}
+	return n, nil
```

Comment:
In practice, Lister interface can only return NotFound error or nil if there is no programing error. I don't mean to skip handling other errors but it's redundant to repeat them in handlers. The wrapped message could be added to controller so it just needs one line in handler, for example:
```
func (c *ClusterGroupHandler) GetNewObject(namespace, name string) (metav1.Object, error) {
        return c.lister.Get(name)
}

newObj, err := c.mirroringHandler.GetNewObject(namespace, name)
if err != nil && !apierrors.IsNotFound(err) {
        return fmt.Errorf("failed to get new %s %s/%s: %v", c.CRDName, namespace, name, err)
}
```

This applies to other methods as well if their error handling are same, which is likely to be true.

Reply:
Got it.